Fix nested int caching for NJT inference tensor

soulitzer opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #132814
* #132414
* __->__ #133061
* #130292
* #131803
* #131916

Tries to solve https://fb.workplace.com/groups/296513499814673/permalink/465746352891386/


cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang @amjames

soulitzer committed:
Fix nested int caching for NJT inference tensor

[ghstack-poisoned]

soulitzer committed:
Update on "Fix nested int caching for NJT inference tensor"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

soulitzer commented on torch/_subclasses/fake_tensor.py:
```diff
@@ -572,9 +583,10 @@ def __set__(self, obj: FakeTensor, value: Optional[torch.SymInt]) -> None:
             setattr(obj, self._memo_vc(obj), None)
             setattr(obj, self._memo_epoch(obj), None)
-        elif not torch.is_inference_mode_enabled():
+        elif not torch.is_inference_mode_enabled() or self._opt_vc:
```
I wonder why we don't just check `obj.is_inference()` here

ezyang replied:
Yeah, is inference seems better

soulitzer committed:
Update on "Fix nested int caching for NJT inference tensor"


Tries to solve https://fb.workplace.com/groups/296513499814673/permalink/465746352891386/


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

ezyang commented:
In general, I don't think you should need an opt_vc. Instead, you check if you have an inference tensor / are in inference mode before you do a vc test. If you are, you skip the vc test.

soulitzer commented:
> In general, I don't think you should need an opt_vc. Instead, you check if you have an inference tensor / are in inference mode before you do a vc test. If you are, you skip the vc test.

Hmm I feel like I don't completely understand. Are you proposing that opt_vc always be True? In my head the point of the vc test was to guard against the case where you call nonzero on x mutate x, then call nonzero on x again. You want the result of the first and second nonzero to have different shapes.

soulitzer committed:
Update on "Fix nested int caching for NJT inference tensor"


Tries to solve https://fb.workplace.com/groups/296513499814673/permalink/465746352891386/


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

soulitzer commented:
Closing as these changes have been added to the reland of the previous PR:  https://github.com/pytorch/pytorch/pull/133196
